reconcile/vault_replication.py

```diff
@@ -0,0 +1,232 @@
+            write_dict = {"path": path, "data": secret}
+            logging.info(["replicate_vault_secret", src_version, path])
+            if not dry_run:
+                dest_vault.write(write_dict)
+        elif dest_version < version:
+            deep_copy_versions(
+                dry_run=dry_run,
+                source_vault=source_vault,
+                dest_vault=dest_vault,
+                current_dest_version=dest_version,
+                current_source_version=version,
+                path=path,
+            )
+        else:
+            logging.info(["replicate_vault_secret", dest_version, version, path])
+    except SecretNotFound:
+        logging.info(["replicate_vault_secret", "Secret not found", path])
+        deep_copy_versions(
+            dry_run=dry_run,
+            source_vault=source_vault,
+            dest_vault=dest_vault,
+            current_dest_version=0,
+            current_source_version=version,
+            path=path,
+        )
+
+
+def check_invalid_paths(
+    path_list: List[str],
+    policy_paths: Optional[List[str]],
+) -> None:
+
+    invalid_paths = []
+    if policy_paths is not None:
+        invalid_paths = list_invalid_paths(path_list, policy_paths)
+        if invalid_paths:
+            logging.error(["replicate_vault_secret", "Invalid paths", invalid_paths])
+            raise VaultInvalidPaths
+
+
+def copy_vault_secrets(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    path_list: List[str],
+) -> None:
+    for path in path_list:
+        copy_vault_secret(dry_run, source_vault, dest_vault, path)
+
+
+def list_invalid_paths(path_list: List[str], policy_paths: List[str]) -> List[str]:
+    invalid_paths = []
+
+    for path in path_list:
+        if not policy_contais_path(path, policy_paths):
+            invalid_paths.append(path)
+
+    return invalid_paths
+
+
+def policy_contais_path(path: str, policy_paths: List[str]) -> bool:
+    return any(path in p_path for p_path in policy_paths)
+
+
+def get_policy_paths(policy_name, instance_name) -> List[str]:
+    query_data = vault_policies.query(query_func=gql.get_api().query)
+    policy_paths = []
+
+    if query_data.policy:
+        for policy in query_data.policy:
+            if policy.name == policy_name and policy.instance.name == instance_name:
+                for line in policy.rules.split("\n"):
+                    res = re.search(r"path \s*[\'\"](.+)[\'\"]", line)
+
+                    if res is not None:
+                        policy_paths.append(res.group(1))
+
+    return policy_paths
+
+
+def get_jenkins_secret_list(jenkins_instance: str) -> List[str]:
+    secret_list = []
+    query_data = jenkins_configs.query(query_func=gql.get_api().query)
+
+    if query_data.jenkins_configs:
+        for p in query_data.jenkins_configs:
+            if (
+                isinstance(p, JenkinsConfigV1_JenkinsConfigV1)
```

Comment:
do we need the `isinstance` bit?

Reply:
Type checking complains if I don't make sure that is a JenkinsConfigV1_JenkinsConfigV1 type, not sure if there is another way to do it